marlowe/src/Language/Marlowe/Core/V1/Semantics/Next.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+
+
+module Language.Marlowe.Core.V1.Semantics.Next
+  ( AmbiguousIntervalProvided(..)
+  , CaseIndex(..)
+  , Next(..)
+  , NextGeneralizedInput(..)
+  , next
+  ) where
+
+import Control.Applicative
+import Control.Monad.Cont ((=<<))
+import Data.Aeson
+import Data.Aeson.Types ()
+import Data.Bifunctor (Bifunctor(first))
+import Data.List.Index (indexed)
+import Data.List.NonEmpty as NEL (NonEmpty(..), nonEmpty, toList)
+import Data.Maybe (catMaybes)
+import Data.Text (pack)
+import Deriving.Aeson (Generic)
+import Language.Marlowe.Core.V1.Semantics
+import Language.Marlowe.Core.V1.Semantics.Types hiding (environment)
+import Language.Marlowe.Pretty (Pretty(..))
+import PlutusTx.Prelude (Bool(..), Either(..), Integer, Maybe(..), maybe, uncurry, ($), (.))
+import qualified Prelude as Haskell
+
+
+
+
+data Next = CanAdvance [NextGeneralizedInput]
+          | CanNotAdvance
+    deriving stock (Haskell.Show)
+
+data AmbiguousIntervalProvided = AmbiguousIntervalProvided
+
+newtype CaseIndex = CaseIndex {unIndex ::Integer}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+    deriving (FromJSON,ToJSON)
+
+-- | Next Ranges of Inputs that can be applied to a given contract
+data NextGeneralizedInput
+  = CanDeposit CaseIndex Party AccountId Token Integer -- 1 possibility > 1 action
+  | CanChoose  CaseIndex ChoiceId [Bound] -- 1 possibility > many different actions
+  | CanNotify  CaseIndex -- 1 possibility > 1 action
+  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+  deriving anyclass (Pretty)
+
+
+next :: Environment ->  State  ->  Contract -> Either AmbiguousIntervalProvided Next
+next  _ _ Close         = Right . CanAdvance $ []
+next  _ _ (When [] _ _) = Right CanNotAdvance   -- Contract Suspended in a time out
+next environment state (When (x:xs) _ _)
+  = Right
+    . maybe
+          CanNotAdvance -- Contract Suspended (only falsified Notifies evaluated in the When)
+        ( CanAdvance . keepFirstNotifyTrueAndRemoveFollowings  )
+    . nextGeneralizedInputsMaybe environment state
+    $ x :| xs
+next environment state  contract  = uncurry (next environment) =<< reduceContract environment state contract
+
+
+keepFirstNotifyTrueAndRemoveFollowings ::  NonEmpty NextGeneralizedInput -> [NextGeneralizedInput]
+keepFirstNotifyTrueAndRemoveFollowings
+  = keepFirstNotifyTrueAndRemoveFollowings' False . NEL.toList
+ where
+  keepFirstNotifyTrueAndRemoveFollowings' :: Bool -> [NextGeneralizedInput] -> [NextGeneralizedInput]
+  keepFirstNotifyTrueAndRemoveFollowings' _ [] =  []
+  keepFirstNotifyTrueAndRemoveFollowings' False (x@CanNotify {} : xs) =  keepFirstNotifyTrueAndRemoveFollowings' True (x:xs)
+  keepFirstNotifyTrueAndRemoveFollowings' True (CanNotify {} : xs) =  keepFirstNotifyTrueAndRemoveFollowings' True xs
+  keepFirstNotifyTrueAndRemoveFollowings' _ xs =  keepFirstNotifyTrueAndRemoveFollowings' True xs
```

Comment:
You could use `nubBy` for this. But actually, I realised you would need to do something similar for `Choices` (and for deposits, but that you could do with `nubBy`). For example, and this is pseudo-haskell:
```haskell
[ Notify FalseObs cont1
, Notify TrueObs cont2
, Deposit 153 cont3
, Choice [(2,5)] cont4
, Choice [(0,7)] cont5
, Deposit 153 cont6
, Deposit 155 cont7
, Notify TrueObs cont8
]
```
Should return
```haskell
[(2, Notify), (3, Deposit 153), (4, Choice [(2,5)]), (5, Choice [(0,1), (6, 7)]), (7, Deposit 155)]
```

It is easier if you don't return the branch position, but still needs some work

Reply:
used a fold in the new version